File: app/src/main/java/com/example/my_board/BoardActivity.java
package com.example.my_board;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.BaseExpandableListAdapter;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.ExpandableListAdapter;
import android.widget.ExpandableListView;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.ToggleButton;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.HashMap;

public class BoardActivity extends AppCompatActivity {
    public ExListViewAdapter adapter;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.board);
        final ExpandableListView listView = (ExpandableListView) findViewById(R.id.listview);
        final FirebaseDatabase database = FirebaseDatabase.getInstance();
        final DatabaseReference Ref = database.getReference("Content/");
        final User user = (User)getApplication();

        final TextView board_title;
        Button Button_list;
        Button Button_delete;
        Button Button_done;
        Button Button_modify;

        TextView board_content;
        final TextView board_comment;
        final TextView countLike;
        final ToggleButton likeImage;

        Button_list = (Button) findViewById(R.id.Button_list);
        Button_done = (Button) findViewById(R.id.Button_done);

        board_title = (TextView)findViewById(R.id.board_title);
        board_content = (TextView)findViewById(R.id.board_content);
        board_comment = (TextView)findViewById(R.id.TextInputEditText_comment);
        likeImage = (ToggleButton)findViewById(R.id.likeImage);
        countLike = (TextView)findViewById(R.id.countLike);

        final Intent intent = getIntent();
        final String board_id = intent.getStringExtra("board_id");
        final String title = intent.getStringExtra("title");
        final String boardUid = intent.getStringExtra("boardUid");
        String content = intent.getStringExtra("content");
        countLike.setText(intent.getStringExtra("countLike"));

        final SharedPreferences preferences = getPreferences(MODE_PRIVATE);
        final DatabaseReference myRef = database.getReference("User/" + user.getUId() + "/likeList");
        likeImage.setTag("1");
        if(Integer.parseInt(likeImage.getTag().toString()) == 1){
            myRef.addValueEventListener(new ValueEventListener() {
                @Override
                public void onDataChange(DataSnapshot dataSnapshot) {
                    if(Integer.parseInt(likeImage.getTag().toString()) == 1){
                        if(dataSnapshot.child(title).getValue() != null){
                            likeImage.setTag("0");
                            likeImage.setChecked(true);
                        }
                    }
                }
                @Override
                public void onCancelled(DatabaseError error) {
                    // Failed to read value
                    Log.w("Failed to read value.", error.toException());
                }
            });
        }


        board_title.setText(title);
        board_content.setText(content);

        Button_list.setOnClickListener(new Button.OnClickListener(){

            @Override
            public void onClick(View v) {
                Intent intent = new Intent(BoardActivity.this, MainActivity.class);
                startActivity(intent);
            }
        });


        Button_done.setOnClickListener(new Button.OnClickListener(){

            @Override
            public void onClick(View v) {
                String comment = board_comment.getText().toString().trim();
                User user = (User)getApplication();
                if(comment.isEmpty()){
                    Toast.makeText(BoardActivity.this, "댓글 작성 실패!", Toast.LENGTH_SHORT).show();
                }
                else{
                    HashMap<Object, String> hashMap = new HashMap<>();
                    hashMap.put("comment", comment);
                    hashMap.put("uid", user.getUId());
                    FirebaseDatabase database = FirebaseDatabase.getInstance();
                    DatabaseReference reference = database.getReference("Content/" + board_id + "/");
                    reference.child("comment/" + user.getUId() + comment + '/').setValue(hashMap);

                    //글 작성 완료 시 가입 화면을 빠져나감.
                    Toast.makeText(BoardActivity.this, "댓글 작성 완료!", Toast.LENGTH_SHORT).show();
                }
            }
        });

        likeImage.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            int boardLike = Integer.parseInt(intent.getStringExtra("countLike"));
            int like = boardLike;
            SharedPreferences preferencesR = getPreferences(MODE_PRIVATE);
            boolean likePrefR = preferencesR.getBoolean(title + "like", false);
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                System.out.println("likeRefR" + likePrefR);
                if(isChecked)
                {
                    if(Integer.parseInt(likeImage.getTag().toString()) == 1) {
                        like++;
                        DatabaseReference likeRef = Ref.child(boardUid + title + "/like/");
                        likeRef.setValue(like);
                        Toast.makeText(BoardActivity.this, "좋아요를 눌렀습니다.", Toast.LENGTH_SHORT).show();
                        countLike.setText(Integer.toString(like));
                        myRef.child(title+"/").setValue(title);
                        SharedPreferences.Editor editor = preferences.edit();
                        editor.putBoolean(title + "like", true); // value to store
                        editor.commit();
                    }
                }
                else
                {
                    like--;
                    likeImage.setChecked(false);
                    DatabaseReference likeRef = Ref.child(boardUid+ title + "/like/");
                    likeRef.setValue(like);
                    Toast.makeText(BoardActivity.this, "좋아요를 취소했습니다.", Toast.LENGTH_SHORT).show();
                    countLike.setText(Integer.toString(like));
                    myRef.child(title+"/").removeValue();
                    SharedPreferences.Editor editor = preferences.edit();
                    editor.putBoolean(title + "like", false); // value to store
                    editor.commit();
                }
            }
        });


        Ref.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                listView.setAdapter((BaseExpandableListAdapter)null);
                adapter = new ExListViewAdapter(null, null, listView);
                adapter.setUId(user.getUId());
                adapter.setBoard_title(board_id);
                int index = 0;

                ArrayList<ListViewItem> parent = new ArrayList<ListViewItem>();
                HashMap<Integer, ArrayList<ListViewItem>> citemList = new HashMap<Integer, ArrayList<ListViewItem>>();
                //parent list와 child list를 연결할 hashmap 변수 선언
                HashMap<ListViewItem, ArrayList<ListViewItem>> childlist = new HashMap<ListViewItem, ArrayList<ListViewItem>>();
                for(DataSnapshot parentSnapshot : dataSnapshot.child(board_id + "/comment").getChildren()){
                    index++;
                    ArrayList<ListViewItem> child= new ArrayList<ListViewItem>();
                    String commentContent = parentSnapshot.child("comment/").getValue().toString();
                    String uid = parentSnapshot.child("uid/").getValue().toString();
                    ListViewItem pitem = new ListViewItem(ContextCompat.getDrawable(BoardActivity.this, R.drawable.icon_notice), commentContent, uid);
                    parent.add(pitem);
                    for(DataSnapshot childSnapshot : parentSnapshot.child("/ccomment").getChildren()){
                        String ccommentContent = childSnapshot.child("comment/").getValue().toString();
                        System.out.println("댓글은 " + commentContent + ", 대댓글은 " + ccommentContent);
                        String cuid = childSnapshot.child("uid/").getValue().toString();
                        ListViewItem citem = new ListViewItem(ContextCompat.getDrawable(BoardActivity.this, R.drawable.icon_notice), ccommentContent, cuid);
                        child.add(citem);
                    }
                    citemList.put(index-1, child);
                    if(!child.isEmpty()){
                        childlist.put(parent.get(index-1), citemList.get(index-1));
                        System.out.println(parent.get(index-1).getBoard_title() + "에 맵핑 완료!!!");
                    }
                }
                adapter.setmParentList(parent);
                adapter.setmChildHashMap(childlist);
                listView.setAdapter(adapter);

                int groupCount = (int)adapter.getGroupCount();
                System.out.println("groundCount" + groupCount);
                for(int i = 0; i < groupCount; i++){
                    listView.expandGroup(i);
                }

            }

            @Override
            public void onCancelled(DatabaseError error) {
                // Failed to read value
                Log.w("Failed to read value.", error.toException());
            }
        });

    }


}

File: app/src/main/java/com/example/my_board/Activity/BoardActivity.kt
package com.example.my_board.Activity

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.example.my_board.*
import com.example.my_board.ListView.ExListViewAdapter
import com.example.my_board.R
import com.google.firebase.database.*
import java.util.*
import kotlinx.android.synthetic.main.board.*

class BoardActivity : AppCompatActivity() {
    var adapter: ExListViewAdapter<*>? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.board)
        val listView = findViewById<View>(R.id.listview) as ExpandableListView
        val database = FirebaseDatabase.getInstance()
        val Ref = database.getReference("Content/")
        val user = application as User
        val intent = intent
        val board_id = intent.getStringExtra("board_id").toString()
        val title = intent.getStringExtra("title")
        val boardUid = intent.getStringExtra("boardUid")
        val content = intent.getStringExtra("content")
        countLike.text = intent.getStringExtra("countLike")
        likeImage.tag = "1"
        if (likeImage.tag.toString().toInt() == 1) {
            Ref.addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if (likeImage.tag.toString().toInt() == 1) {
                        if (dataSnapshot.child(boardUid + title + "/like/" + user.uId).getValue() != null) {
                            likeImage.tag = "0"
                            likeImage.isChecked = true
                            likeImage.tag = "2"
                        }
                    }
                }

                override fun onCancelled(error: DatabaseError) {
                    // Failed to read value
                    Log.w("Failed to read value.", error.toException())
                }
            })
        }
        board_title.text = title
        board_content.text = content
        Button_list.setOnClickListener {
            val intent = Intent(this@BoardActivity, MainActivity::class.java)
            startActivity(intent)
        }
        Button_done.setOnClickListener {
            val comment = TextInputEditText_comment.text.toString().trim { it <= ' ' }
            val user = application as User
            if (comment.isEmpty()) {
                Toast.makeText(this@BoardActivity, "댓글 작성 실패!", Toast.LENGTH_SHORT).show()
            } else {
                val hashMap = HashMap<Any, String>()
                hashMap["comment"] = comment
                hashMap["uid"] = user.uId!!
                val database = FirebaseDatabase.getInstance()
                val reference = database.getReference("Content/$board_id/")
                reference.child("comment/" + user.uId + comment + '/').setValue(hashMap)

                //글 작성 완료 시 가입 화면을 빠져나감.
                Toast.makeText(this@BoardActivity, "댓글 작성 완료!", Toast.LENGTH_SHORT).show()
            }
        }
        likeImage.setOnCheckedChangeListener(object : CompoundButton.OnCheckedChangeListener {
            var boardLike = intent.getStringExtra("countLike").toInt()
            var like = boardLike
            override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
                if (isChecked) {
                    if (likeImage.tag.toString().toInt() == 1 || likeImage.tag.toString().toInt() == 2) {
                        like++
                        val likeRef = Ref.child(boardUid + title + "/like/" + user.uId)
                        likeRef.setValue(user.uId)
                        Toast.makeText(this@BoardActivity, "좋아요를 눌렀습니다.", Toast.LENGTH_SHORT).show()
                        countLike.text = Integer.toString(like)
                    }
                } else {
                    like--
                    val likeRef = Ref.child(boardUid + title + "/like/" + user.uId)
                    likeRef.removeValue()
                    Toast.makeText(this@BoardActivity, "좋아요를 취소했습니다.", Toast.LENGTH_SHORT).show()
                    countLike.text = Integer.toString(like)
                }
            }
        })
        Ref.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                listView.setAdapter(null as BaseExpandableListAdapter?)
                val childlist = HashMap<ListViewItem, ArrayList<ListViewItem>?>()
                val parent = ArrayList<ListViewItem>()
                val adapter = ExListViewAdapter<Any?>(parent, childlist, listView)
                adapter.setUId(user.uId)
                adapter.setBoard_title(board_id)
                var index = 0
                val citemList = HashMap<Int, ArrayList<ListViewItem>>()
                for (parentSnapshot in dataSnapshot.child("$board_id/comment").children) {
                    index++
                    val child = ArrayList<ListViewItem>()
                    val commentContent = parentSnapshot.child("comment/").value.toString()
                    val uid = parentSnapshot.child("uid/").value.toString()
                    val pitem = ListViewItem(ContextCompat.getDrawable(this@BoardActivity, R.drawable.icon_notice), commentContent, uid)
                    parent.add(pitem)
                    for (childSnapshot in parentSnapshot.child("/ccomment").children) {
                        val ccommentContent = childSnapshot.child("comment/").value.toString()
                        println("댓글은 $commentContent, 대댓글은 $ccommentContent")
                        val cuid = childSnapshot.child("uid/").value.toString()
                        val citem = ListViewItem(ContextCompat.getDrawable(this@BoardActivity, R.drawable.icon_notice), ccommentContent, cuid)
                        child.add(citem)
                    }
                    citemList[index - 1] = child
                    if (!child.isEmpty()) {
                        childlist[parent[index - 1]] = citemList[index - 1]!!
                        println(parent[index - 1].board_title + "에 맵핑 완료!!!")
                    }
                }
                adapter.setmParentList(parent)
                adapter.setmChildHashMap(childlist)
                listView.setAdapter(adapter)
                val groupCount = adapter.getGroupCount()
                println("groundCount$groupCount")
                for (i in 0 until groupCount) {
                    listView.expandGroup(i)
                }
            }

            override fun onCancelled(error: DatabaseError) {
                // Failed to read value
                Log.w("Failed to read value.", error.toException())
            }
        })
    }
}